### DISCUSSION TITLE
Issue with Path Override Warning and Markdown Preview in `qutebrowser`

### DISCUSSION THREAD

**Discussion by Aman1337g** [qutebrowser/qutebrowser#8385] [Q&A] [answered]
Hello,

I'm encountering an issue with `qutebrowser` and the [Markdown preview](https://github.com/iamcco/markdown-preview.nvim) integration when using Neovim’s `:MarkdownPreview` command. Specifically, I’m getting a path override warning related to missing dictionary files during the initial launch of `qutebrowser`, but only when no instance of `qutebrowser` is already running.

### The Problem:
- When I launch `qutebrowser` **for the first time** from the virtual environment, I receive the following warning:
  ```
  WARNING: Path override failed for key base::DIR_APP_DICTIONARIES and path '/home/aman/qutebrowser/.venv/bin/qtwebengine_dictionaries'
  ```
  This warning seems to be related to missing dictionary files required by the `QtWebEngine` component.
  
- However, if there is an **existing instance** of `qutebrowser` running, this warning does **not** appear, and the Markdown preview works correctly in `qutebrowser` when triggered by the `:MarkdownPreview` command in Neovim. The message logged in this case is:
  ```
  INFO: Opening in existing instance of qutebrowser http://localhost:8177/page/1
  ```

### Steps to Reproduce:
1. Launch `qutebrowser` from the virtual environment (without any existing instance running).
2. Trigger a Markdown preview in Neovim using `:MarkdownPreview`.
3. Observe the warning about path override failure and the Markdown preview not being displayed correctly.

### Observations:
- The warning only appears when starting a **new instance** of `qutebrowser`.
- The Markdown preview is displayed correctly if an existing instance of `qutebrowser` is already running.
  
### What I’ve Tried:
1. Reinstalling `PyQt6` and `pyqt6-webengine` with the `--force-reinstall` flag.
2. Verifying that the necessary dictionary files are missing in the virtual environment's `libexec` directory.
3. Searching for any missing dictionary files, but no additional files were found.

### Question:
Is there any specific configuration or workaround that can ensure `qutebrowser` properly loads its dictionaries even when started for the first time from a virtual environment? Any suggestions on how to resolve this path override warning would be greatly appreciated.

### Markdown Preview Configuration:

I am using the `iamcco/markdown-preview.nvim` plugin in Neovim for live Markdown preview. Below is the configuration I have set up for it:

```lua
{
  "iamcco/markdown-preview.nvim",
  cmd = { "MarkdownPreview", "MarkdownPreviewStop", "MarkdownPreviewToggle" },
  ft = { "markdown" },
  build = function()
    vim.fn["mkdp#util#install"]()
  end,
  config = function()
    -- Auto-start preview
    vim.g.mkdp_auto_start = 0
    vim.g.mkdp_auto_close = 1
    vim.g.mkdp_refresh_slow = 0
    vim.g.mkdp_command_for_global = 0
    vim.g.mkdp_open_to_the_world = 0
    vim.g.mkdp_open_ip = ""
    vim.g.mkdp_echo_preview_url = 0

    -- Use qutebrowser as the preview browser
    vim.g.mkdp_browserfunc = "MkdpBrowserFn"
    local cmd = "/home/aman/qutebrowser/.venv/bin/python3 -m qutebrowser"
    vim.api.nvim_exec2(
      string.format(
        [[
          function! MkdpBrowserFn(url)
              execute '!%s' a:url
          endfunction
        ]],
        cmd
      ),
      {}
    )

    -- Preview options
    vim.g.mkdp_preview_options = {
      mkit = {},
      katex = {},
      uml = {},
      maid = {},
      disable_sync_scroll = 0,
      sync_scroll_type = "middle",
      hide_yaml_meta = 1,
      sequence_diagrams = {},
      flowchart_diagrams = {},
      content_editable = false,
      disable_filename = 0,
      toc = {},
    }

    -- Optional settings
    vim.g.mkdp_markdown_css = ""
    vim.g.mkdp_highlight_css = ""
    vim.g.mkdp_port = "8177"
    vim.g.mkdp_page_title = "「${name}」"
    vim.g.mkdp_filetypes = { "markdown" }
    vim.g.mkdp_theme = "dark"
    vim.g.mkdp_combine_preview = 0
    vim.g.mkdp_combine_preview_auto_refresh = 1
  end,
}
```

The plugin is configured to use `qutebrowser` as the browser for rendering the Markdown preview, but I am facing issues when `qutebrowser` is started without an existing instance.

---

Thank you in advance for any help or suggestions on resolving the path override warning and ensuring that the Markdown preview works consistently with `qutebrowser`.

**Comment by The-Compiler** [comment]
The warning is coming from QtWebEngine 6.8 - it's harmless and unrelated, it just says that no spell checking is configured. You don't seem to say anywhere what the *actual* issue really is? "not being displayed correctly" how exactly?

**Reply by Aman1337g** [reply]
Thank you for your response! Let me clarify the issue further.

### What Happens:
1. When `:MarkdownPreview` is executed in Neovim, `qutebrowser` launches (if it’s not already running).
2. However, the preview page (e.g., `http://localhost:8177/page/1`) initially displays as a **black screen** in `qutebrowser`.
3. If I press `Ctrl+C` in Neovim:
- `qutebrowser` **closes completely**, and the Markdown preview content no longer has a browser instance to display it.

### Observations:
- **When `qutebrowser` is already running**: The preview loads immediately without issue, with no black screen.
- **When `qutebrowser` is not running**: The black screen issue occurs, and pressing `Ctrl+C` in Neovim closes the newly launched `qutebrowser` instance entirely.

### What I’ve Tried:
- Confirmed the `qutebrowser` instance is being opened correctly with:
  ```
  INFO: Opening in existing instance of qutebrowser http://localhost:8177/page/1
  ```
- Verified that no error is logged in Neovim after pressing `Ctrl+C`.
- The QtWebEngine warning about `DIR_APP_DICTIONARIES` does not seem to affect functionality as you mentioned, so it’s unrelated to this behavior.

### My Question:
Could this behavior be caused by:
- A timing issue, where the `MarkdownPreview` command exits before the browser fully initializes?
- Some kind of miscommunication between `Neovim`'s `markdown-preview.nvim` plugin and `qutebrowser` when launching a new instance?

### My Current Workaround:
Starting `qutebrowser` manually before running `:MarkdownPreview` avoids the issue, but I’d like to understand what causes the initial black screen when launching `qutebrowser` programmatically.

**Reply by The-Compiler** [reply]
Probably the way you're launching the browser from vim is blocking, but it shouldn't be. Have you tried with a different browser that blocks when invoked the first time too (I believe Chromium does)?

**Reply by Aman1337g** [reply, marked as answer]
I checked with Chromium-based Brave Browser and encountered the same issue. However, after using the following approach, it worked:

```vim
vim.g.mkdp_browserfunc = "MkdpBrowserFn"
vim.api.nvim_exec2(
  [[
  function! MkdpBrowserFn(url)
      call jobstart(['brave-browser', a:url])
  endfunction
  ]],
  {}
)
```

This resolved the issue for Brave. Similarly, for qutebrowser, I used:

```vim
vim.g.mkdp_browserfunc = "MkdpQutebrowserFn"
vim.api.nvim_exec2(
  [[
  function! MkdpQutebrowserFn(url)
      " Launch qutebrowser in a non-blocking way
     call jobstart(['/home/aman/qutebrowser/.venv/bin/python3', '-m', 'qutebrowser', a:url])
  endfunction
  ]],
  {}
)
```

Both browsers now open as expected without blocking. Thanks for the suggestions!
